packages/make-hardener/src/main.js

```diff
@@ -33,113 +59,216 @@ const { ownKeys } = Reflect;
-        // Object are verified before being enqueued,
-        // therefore this is a valid candidate.
-        // Throws if this fails (strict mode).
-        freeze(obj);
-
-        // we rely upon certain commitments of Object.freeze and proxies here
-
-        // get stable/immutable outbound links before a Proxy has a chance to do
-        // something sneaky.
-        const path = paths.get(obj) || 'unknown';
-        const descs = getOwnPropertyDescriptors(obj);
-        const proto = getPrototypeOf(obj);
-        enqueue(proto, `${path}.__proto__`);
-
-        ownKeys(descs).forEach(name => {
-          const pathname = `${path}.${String(name)}`;
-          // todo uncurried form
-          // todo: getOwnPropertyDescriptors is guaranteed to return well-formed
-          // descriptors, but they still inherit from Object.prototype. If
-          // someone has poisoned Object.prototype to add 'value' or 'get'
-          // properties, then a simple 'if ("value" in desc)' or 'desc.value'
-          // test could be confused. We use hasOwnProperty to be sure about
-          // whether 'value' is present or not, which tells us for sure that
-          // this is a data property.
-          // The 'name' may be a symbol, and TypeScript doesn't like us to
-          // index arbitrary symbols on objects, so we pretend they're just
-          // strings.
-          const desc = descs[/** @type {string} */ (name)];
-          if ('value' in desc) {
-            // todo uncurried form
-            enqueue(desc.value, `${pathname}`);
-          } else {
-            enqueue(desc.get, `${pathname}(get)`);
-            enqueue(desc.set, `${pathname}(set)`);
-          }
-        });
+  const patchReflectMethods = {
+    // Note that Object.preventExtensions has a slightly different
+    // semantics than Reflect.preventExtensions
+    preventExtensions(target) {
+      if (isExtensible(target)) {
+        onlyNonExtensible.add(target);
       }
+      return originalReflectPreventExtensions(target);
+    },
+  };
+
+  if (ENABLE_ONLY_NON_EXTENSIBLE) {
+    defineProperties(
+      Object,
+      Object.getOwnPropertyDescriptors(patchObjectMethods),
+    );
+
+    defineProperties(
+      Reflect,
+      Object.getOwnPropertyDescriptors(patchReflectMethods),
+    );
+  }
+
+  /**
+   * if ENABLE_ONLY_NON_EXTENSIBLE then this will be the
+   * patched one. Otherwise it will be the original.
+   */
+  const { freeze: patchedObjectFreeze } = Object;
 
-      function dequeue() {
-        // New values added before forEach() has finished will be visited.
-        toFreeze.forEach(freezeAndTraverse); // todo curried forEach
+  /**
+   * Currently not exported. But we should consider exporting it
+   * as a SES peer of isFrozen, etc. Of course, we'd need to remove
+   * the `path` parameter first.
+   *
+   * @param {any} val
+   * @param {string=} path
+   * @returns {boolean}
+   */
+  const isHardened = (val, path) => {
+    if (val !== Object(val)) {
```

Comment:
Since we are trying to reduce the memory pressure due to harden with this change, is there a way to make this test that doesn't require an allocation?  e.g., `typeof val !== 'object' && typeof val !== 'function'`

Reply:
I added an `isPrimitive` function so I can switch these to use that. But they are not yet switched, so I'm leaving this conversation unresolved for now. Btw, your code snippet above forgot to check for `null`, and I did when I first wrote `isPrimitive`.

Reply:
done. resolving.